src/main/java/graphql/validation/rules/SingleSubscriptionFieldOnly.java

```diff
@@ -0,0 +1,51 @@
+package graphql.validation.rules;
+
+import graphql.Internal;
+import graphql.language.OperationDefinition;
+import graphql.language.Selection;
+import graphql.validation.AbstractRule;
+import graphql.validation.ValidationContext;
+import graphql.validation.ValidationErrorCollector;
+import graphql.validation.ValidationErrorType;
+
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A GraphQL document is only valid if each subscription operation has exactly one root field
+ * https://spec.graphql.org/June2018/#sec-Single-root-field
+ */
+@Internal
+public class SingleSubscriptionFieldOnly extends AbstractRule {
+
+    private Set<String> operationNames = new LinkedHashSet<>();
+
+    public SingleSubscriptionFieldOnly(ValidationContext validationContext, ValidationErrorCollector validationErrorCollector) {
+        super(validationContext, validationErrorCollector);
+    }
+
+    @Override
+    public void checkOperationDefinition(OperationDefinition operationDefinition) {
+        super.checkOperationDefinition(operationDefinition);
+
+        // skip validation for non subscription operations
+        if (!isSubscriptionOperation(operationDefinition)) {
+            return;
+        }
+
+        List<Selection> selections = operationDefinition.getSelectionSet().getSelections();
+        if (selections.size() > 1) {
```

Comment:
Close but not quite.

This needs to be the output of CollectFields - eg `graphql.execution.FieldCollector`

However I am not sure we can do that in validation rules can we??

I checked and there is no code using `graphql.execution.FieldCollector` today in validation

Is it ok to do so?

If you look at the spec there give and example of collected fields

Reply:
Single root field

Formal Specification

For each subscription operation definition subscription in the document
* Let subscriptionType be the root Subscription type in schema.
* Let selectionSet be the top level selection set on subscription.
* Let variableValues be the empty set.
* Let groupedFieldSet be the result of `CollectFields(subscriptionType, selectionSet, variableValues)`.
* groupedFieldSet must have exactly one entry.

Reply:
the FieldCollector methods require a lot more parameters than what is available in validation, which just traverses through the nodes, will have to see if it is even possible to provide the parameters, when in the context of validation we only have the Node object, and not other things like schema

Reply:
See `graphql.validation.ValidationContext` - every validation rule has it.  If we need more (I think we need variables) then we should extend it to contain what we need

The schema and fragments are all present